libr/core/canal.c

```diff
@@ -2753,39 +2753,90 @@ R_API void r_core_recover_vars(RCore *core, RAnalFunction *fcn, bool argonly) {
 	return;
 }
 
-R_API RList* r_core_anal_graph_to(RCore *core, ut64 addr, int n) {
-	RAnalBlock *bb, *root = NULL, *dest = NULL;
-	RListIter *iter, *iter2;
-	RList *list2 = NULL, *list = NULL;
-	RAnalFunction *fcn;
+static int cmp_addr(const void *a, const void *b) {
+	const ut64 *_a, *_b;
+
+	_a = a;
+	_b = b;
+
+	return *_a - *_b;
+}
+
+static bool anal_path_exists(RCore *core, ut64 from, ut64 to, RList *bbs, int depth, RList *state) {
+	RAnalBlock *bb = r_anal_bb_from_offset (core->anal, from);
+	RListIter *iter = NULL;
+	RList *refs = NULL;
+	RAnalRef *refi;
+	RAnalFunction *fcn = NULL;
+	RAnalFunction *cur_fcn = NULL;
+	ut64 *frp = R_NEW0 (ut64);
+
+	if (depth < 1) {
+		eprintf ("going too deep\n");
+		return false;
+	}
+
+	if (!bb || !bbs) {
+		return false;
+	}
+
+	*frp = from;
+	r_list_append (state, (void *)frp);
+
+	ut64 tj = bb->jump;
+	ut64 fj = bb->addr + bb->size;
+
+	// try to find the target in the current function
+	if (r_anal_bb_is_in_offset (bb, to) ||
+		((!r_list_find (state, (void *)&(tj), cmp_addr) && anal_path_exists (core, bb->jump, to, bbs, depth - 1, state))) ||
+		((!r_list_find (state, (void *)&(fj), cmp_addr) && anal_path_exists (core, bb->addr + bb->size, to, bbs, depth - 1, state)))) {
+		r_list_prepend (bbs, bb);
+		return true;
+	}
 
+	// find our current function
 	r_list_foreach (core->anal->fcns, iter, fcn) {
-		if (!r_anal_fcn_is_in_offset (fcn, core->offset)) {
-			continue;
+		if (r_anal_fcn_in (fcn, from)) {
+			cur_fcn = fcn;
+			break;
 		}
-		r_list_foreach (fcn->bbs, iter2, bb) {
-			if (r_anal_bb_is_in_offset (bb, addr)) {
-				dest = bb;
-			}
-			if (r_anal_bb_is_in_offset (bb, core->offset)) {
-				root = bb;
-				r_list_append (list, list2);
+	}
+
+	// get call refs from current basic block and find a path from them
+	if (cur_fcn) {
+		refs = r_anal_fcn_get_refs (core->anal, cur_fcn);
+		if (refs) {
+			r_list_foreach (refs, iter, refi) {
+				if (refi->type == R_ANAL_REF_TYPE_CALL) {
+					if (r_anal_bb_is_in_offset (bb, refi->at)) {
+						if (refi->at != refi->addr && !r_list_find (state, (void *)&(refi->addr), cmp_addr) && anal_path_exists (core, refi->addr, to, bbs, depth - 1, state)) {
+							r_list_prepend (bbs, bb);
+							return true;
+						}
+					}
+				}
 			}
 		}
 	}
-	if (root && dest) {
-		if (dest == root) {
-			eprintf ("Source and destination are the same\n");
-			return NULL;
-		}
-		eprintf ("ROOT BB 0x%08"PFMT64x"\n", root->addr);
-		eprintf ("DEST BB 0x%08"PFMT64x"\n", dest->addr);
-		list = r_list_new ();
-		printf ("=>  0x%08"PFMT64x"\n", root->jump);
-	} else {
-		eprintf ("Unable to find source or destination basic block\n");
+
+	return false;
+}
+
+R_API RList* r_core_anal_graph_to(RCore *core, ut64 addr, int n) {
+	RList *list = r_list_new ();
+	RList *state = r_list_new ();
+
+	r_return_val_if_fail (list && state, NULL);
```

Comment:
nope! You should just handle the possible NULL values as usual. r_return_ functions should be used for programming errors only, not for runtime error conditions. If `r_list_new()` returns NULL that's probably because we are out of memory and that is something the programmer cannot do anything about, so it's not his fault.

As written in `r_assert.h`
```
 * If @expr evaluates to %FALSE, the current function should be considered to
 * have undefined behaviour (a programmer error). The only correct solution
 * to such an error is to change the module that is calling the current
 * function, so that it avoids this incorrect call.
```

Reply:
done